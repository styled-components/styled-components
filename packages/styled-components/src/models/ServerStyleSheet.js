// @flow
/* eslint-disable no-underscore-dangle */
import React from 'react';
import { IS_BROWSER, SC_ATTR, SC_ATTR_VERSION, SC_VERSION } from '../constants';
import throwStyledError from '../utils/error';
import getNonce from '../utils/nonce';
import StyleSheet from '../sheet';
import StyleSheetManager from './StyleSheetManager';

declare var __SERVER__: boolean;

const CLOSING_TAG_R = /^\s*<\/[a-z]/i;

export default class ServerStyleSheet {
  isStreaming: boolean;

  instance: StyleSheet;

  sealed: boolean;

  constructor() {
    this.instance = new StyleSheet({ isServer: true });
    this.sealed = false;
  }

  _emitSheetCSS = (): string => {
    const css = this.instance.toString();
    if (!css) return '';

    const nonce = getNonce();
    const attrs = [nonce && `nonce="${nonce}"`, `${SC_ATTR}="true"`, `${SC_ATTR_VERSION}="${SC_VERSION}"`];
    const htmlAttr = attrs.filter(Boolean).join(' ');

    return `<style ${htmlAttr}>${css}</style>`;
  };

  collectStyles(children: any) {
    if (this.sealed) {
      return throwStyledError(2);
    }

    return <StyleSheetManager sheet={this.instance}>{children}</StyleSheetManager>;
  }

  getStyleTags = (): string => {
    if (this.sealed) {
      return throwStyledError(2);
    }

    return this._emitSheetCSS();
  };

  getStyleElement = () => {
    if (this.sealed) {
      return throwStyledError(2);
    }

    const props = {
      [SC_ATTR]: '',
      [SC_ATTR_VERSION]: SC_VERSION,
      dangerouslySetInnerHTML: {
        __html: this.instance.toString(),
      },
    };

    const nonce = getNonce();
    if (nonce) {
      (props: any).nonce = nonce;
    }

    // v4 returned an array for this fn, so we'll do the same for v5 for backward compat
    return [<style {...props} key="sc-0-0" />];
  };

  // eslint-disable-next-line consistent-return
  interleaveWithNodeStream(input: any) {
    if (!__SERVER__ || IS_BROWSER) {
      return throwStyledError(3);
    } else if (this.sealed) {
      return throwStyledError(2);
    }

    if (__SERVER__) {
      this.seal();

      // eslint-disable-next-line global-require
      const { Readable, Transform } = require('stream');

      const readableStream: Readable = input;
      const { instance: sheet, _emitSheetCSS } = this;

      const transformer = new Transform({
        transform: function appendStyleChunks(chunk, /* encoding */ _, callback) {
          // Get the chunk and retrieve the sheet's CSS as an HTML chunk,
          // then reset its rules so we get only new ones for the next chunk
          const renderedHtml = chunk.toString();
          const html = _emitSheetCSS();

          sheet.clearTag();

          // prepend style html to chunk, unless the start of the chunk is a
          // closing tag in which case append right after that
          if (CLOSING_TAG_R.test(renderedHtml)) {
            const endOfClosingTag = renderedHtml.indexOf('>') + 1;
            const before = renderedHtml.slice(0, endOfClosingTag);
            const after = renderedHtml.slice(endOfClosingTag);

            this.push(before + html + after);
          } else {
            this.push(html + renderedHtml);
          }

          callback();
        },
      });

      readableStream.on('error', err => {
        // forward the error to the transform stream
        transformer.emit('error', err);
      });

      return readableStream.pipe(transformer);
    }
  }

  seal = () => {
    this.sealed = true;
  };
}
