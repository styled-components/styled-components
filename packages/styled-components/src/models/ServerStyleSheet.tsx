/* eslint-disable no-underscore-dangle */
import React from 'react';
import type * as streamInternal from 'stream';
import { Readable } from 'stream';
import { IS_BROWSER, SC_ATTR, SC_ATTR_VERSION, SC_VERSION } from '../constants';
import StyleSheet from '../sheet';
import styledError from '../utils/error';
import getNonce from '../utils/nonce';
import StyleSheetManager from './StyleSheetManager';

declare const __SERVER__: boolean;

const CLOSING_TAG_R = /^\s*<\/[a-z]/i;

export default class ServerStyleSheet {
  instance: StyleSheet;
  sealed: boolean;

  constructor() {
    this.instance = new StyleSheet({ isServer: true });
    this.sealed = false;
  }

  _emitSheetCSS = (): string => {
    const css = this.instance.toString();
    const nonce = getNonce();
    const attrs = [
      nonce && `nonce="${nonce}"`,
      `${SC_ATTR}="true"`,
      `${SC_ATTR_VERSION}="${SC_VERSION}"`,
    ];
    const htmlAttr = attrs.filter(Boolean).join(' ');

    return `<style ${htmlAttr}>${css}</style>`;
  };

  collectStyles(children: any): JSX.Element {
    if (this.sealed) {
      throw styledError(2);
    }

    return <StyleSheetManager sheet={this.instance}>{children}</StyleSheetManager>;
  }

  getStyleTags = (): string => {
    if (this.sealed) {
      throw styledError(2);
    }

    return this._emitSheetCSS();
  };

  getStyleElement = () => {
    if (this.sealed) {
      throw styledError(2);
    }

    const props = {
      [SC_ATTR]: '',
      [SC_ATTR_VERSION]: SC_VERSION,
      dangerouslySetInnerHTML: {
        __html: this.instance.toString(),
      },
    };

    const nonce = getNonce();
    if (nonce) {
      (props as any).nonce = nonce;
    }

    // v4 returned an array for this fn, so we'll do the same for v5 for backward compat
    return [<style {...props} key="sc-0-0" />];
  };

  // eslint-disable-next-line consistent-return
  // @ts-expect-error alternate return types are not possible due to code transformation
  interleaveWithNodeStream(input: Readable): streamInternal.Transform {
    if (!__SERVER__ || IS_BROWSER) {
      throw styledError(3);
    } else if (this.sealed) {
      throw styledError(2);
    }

    if (__SERVER__) {
      this.seal();

      // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
      const { Transform } = require('stream');

      const readableStream: Readable = input;
      const { instance: sheet, _emitSheetCSS } = this;

      const transformer: streamInternal.Transform = new Transform({
        transform: function appendStyleChunks(
          chunk: string,
          /* encoding */
          _: string,
          callback: Function
        ) {
          // Get the chunk and retrieve the sheet's CSS as an HTML chunk,
          // then reset its rules so we get only new ones for the next chunk
          const renderedHtml = chunk.toString();
          const html = _emitSheetCSS();

          sheet.clearTag();

          // prepend style html to chunk, unless the start of the chunk is a
          // closing tag in which case append right after that
          if (CLOSING_TAG_R.test(renderedHtml)) {
            const endOfClosingTag = renderedHtml.indexOf('>') + 1;
            const before = renderedHtml.slice(0, endOfClosingTag);
            const after = renderedHtml.slice(endOfClosingTag);

            this.push(before + html + after);
          } else {
            this.push(html + renderedHtml);
          }

          callback();
        },
      });

      readableStream.on('error', err => {
        // forward the error to the transform stream
        transformer.emit('error', err);
      });

      return readableStream.pipe(transformer);
    }
  }

  seal = (): void => {
    this.sealed = true;
  };
}
