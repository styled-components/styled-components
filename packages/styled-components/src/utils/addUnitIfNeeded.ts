import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function addUnitIfNeeded(name: string, value: any) {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  if (value == null || typeof value === 'boolean' || value === '') {
    return '';
  }

  if (typeof value === 'number' && value !== 0 && !(name in unitless)) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
