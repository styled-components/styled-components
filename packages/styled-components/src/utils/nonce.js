// @flow
/* eslint-disable camelcase, no-undef */

declare var __webpack_nonce__: string;

const getNonce = () => {
  return typeof __webpack_nonce__ !== 'undefined' ? __webpack_nonce__ : null;
};

export default getNonce;
