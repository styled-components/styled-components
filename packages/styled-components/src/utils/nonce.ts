declare let __webpack_nonce__: string;

export default function getNonce() {
  return typeof __webpack_nonce__ !== 'undefined' ? __webpack_nonce__ : null;
}
