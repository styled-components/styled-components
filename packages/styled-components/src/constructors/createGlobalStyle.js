// @flow
import React, { useContext, useLayoutEffect, useRef } from 'react';
import { STATIC_EXECUTION_CONTEXT } from '../constants';
import GlobalStyle from '../models/GlobalStyle';
import { useStyleSheet, useStylis } from '../models/StyleSheetManager';
import { ThemeContext } from '../models/ThemeProvider';
import type { Interpolation } from '../types';
import { checkDynamicCreation } from '../utils/checkDynamicCreation';
import determineTheme from '../utils/determineTheme';
import generateComponentId from '../utils/generateComponentId';
import css from './css';

declare var __SERVER__: boolean;

type GlobalStyleComponentPropsType = Object;

export default function createGlobalStyle(
  strings: Array<string>,
  ...interpolations: Array<Interpolation>
) {
  const rules = css(strings, ...interpolations);
  const styledComponentId = `sc-global-${generateComponentId(JSON.stringify(rules))}`;
  const globalStyle = new GlobalStyle(rules, styledComponentId);

  if (process.env.NODE_ENV !== 'production') {
    checkDynamicCreation(styledComponentId);
  }

  function GlobalStyleComponent(props: GlobalStyleComponentPropsType) {
    const styleSheet = useStyleSheet();
    const stylis = useStylis();
    const theme = useContext(ThemeContext);
    const instanceRef = useRef(styleSheet.allocateGSInstance(styledComponentId));

    const instance = instanceRef.current;

    if (process.env.NODE_ENV !== 'production' && React.Children.count(props.children)) {
      // eslint-disable-next-line no-console
      console.warn(
        `The global style component ${styledComponentId} was given child JSX. createGlobalStyle does not render children.`
      );
    }

    if (
      process.env.NODE_ENV !== 'production' &&
      rules.some(rule => typeof rule === 'string' && rule.indexOf('@import') !== -1)
    ) {
      // eslint-disable-next-line no-console
      console.warn(
        `Please do not use @import CSS syntax in createGlobalStyle at this time, as the CSSOM APIs we use in production do not handle it well. Instead, we recommend using a library such as react-helmet to inject a typical <link> meta tag to the stylesheet, or simply embedding it manually in your index.html <head> section for a simpler app.`
      );
    }

    if (styleSheet.server) {
      renderStyles(instance, props, styleSheet, theme, stylis);
    }

    if (!__SERVER__) {
      // this conditional is fine because it is compiled away for the relevant builds during minification,
      // resulting in a single unguarded hook call
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useLayoutEffect(() => {
        if (!styleSheet.server) {
          renderStyles(instance, props, styleSheet, theme, stylis);
          return () => globalStyle.removeStyles(instance, styleSheet);
        }
      }, [instance, props, styleSheet, theme, stylis]);
    }

    return null;
  }

  function renderStyles(instance, props, styleSheet, theme, stylis) {
    if (globalStyle.isStatic) {
      globalStyle.renderStyles(instance, STATIC_EXECUTION_CONTEXT, styleSheet, stylis);
    } else {
      const context = {
        ...props,
        theme: determineTheme(props, theme, GlobalStyleComponent.defaultProps),
      };

      globalStyle.renderStyles(instance, context, styleSheet, stylis);
    }
  }

  // $FlowFixMe
  return React.memo(GlobalStyleComponent);
}
