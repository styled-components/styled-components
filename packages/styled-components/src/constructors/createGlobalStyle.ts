import React, { useContext, useLayoutEffect, useRef } from 'react';
import { STATIC_EXECUTION_CONTEXT } from '../constants';
import GlobalStyle from '../models/GlobalStyle';
import { useStyleSheet, useStylis } from '../models/StyleSheetManager';
import { Theme, ThemeContext } from '../models/ThemeProvider';
import StyleSheet from '../sheet';
import { ExtensibleObject, Interpolation, RuleSet, Stringifier, Styles } from '../types';
import { checkDynamicCreation } from '../utils/checkDynamicCreation';
import determineTheme from '../utils/determineTheme';
import generateComponentId from '../utils/generateComponentId';
import css from './css';

declare const __SERVER__: boolean;

export default function createGlobalStyle(
  strings: Styles,
  ...interpolations: Array<Interpolation>
) {
  const rules = css(strings, ...interpolations) as RuleSet;
  const styledComponentId = `sc-global-${generateComponentId(JSON.stringify(rules))}`;
  const globalStyle = new GlobalStyle(rules, styledComponentId);

  if (process.env.NODE_ENV !== 'production') {
    checkDynamicCreation(styledComponentId);
  }

  const GlobalStyleComponent: React.ComponentType<ExtensibleObject> = props => {
    const styleSheet = useStyleSheet();
    const stylis = useStylis();
    const theme = useContext(ThemeContext);
    const instanceRef = useRef(styleSheet.allocateGSInstance(styledComponentId));

    const instance = instanceRef.current;

    if (process.env.NODE_ENV !== 'production' && React.Children.count(props.children)) {
      // eslint-disable-next-line no-console
      console.warn(
        `The global style component ${styledComponentId} was given child JSX. createGlobalStyle does not render children.`
      );
    }

    if (
      process.env.NODE_ENV !== 'production' &&
      rules.some(rule => typeof rule === 'string' && rule.indexOf('@import') !== -1)
    ) {
      // eslint-disable-next-line no-console
      console.warn(
        `Please do not use @import CSS syntax in createGlobalStyle at this time, as the CSSOM APIs we use in production do not handle it well. Instead, we recommend using a library such as react-helmet to inject a typical <link> meta tag to the stylesheet, or simply embedding it manually in your index.html <head> section for a simpler app.`
      );
    }

    if (__SERVER__) {
      renderStyles(instance, props, styleSheet, theme, stylis);
    } else {
      // this conditional is fine because it is compiled away for the relevant builds during minification,
      // resulting in a single unguarded hook call
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useLayoutEffect(() => {
        renderStyles(instance, props, styleSheet, theme, stylis);
        return () => globalStyle.removeStyles(instance, styleSheet);
      }, [instance, props, styleSheet, theme, stylis]);
    }

    return null;
  };

  function renderStyles(
    instance: number,
    props: ExtensibleObject,
    styleSheet: StyleSheet,
    theme: Theme | undefined,
    stylis: Stringifier
  ) {
    if (globalStyle.isStatic) {
      globalStyle.renderStyles(instance, STATIC_EXECUTION_CONTEXT, styleSheet, stylis);
    } else {
      const context = {
        ...props,
        theme: determineTheme(props, theme, GlobalStyleComponent.defaultProps),
      };

      globalStyle.renderStyles(instance, context, styleSheet, stylis);
    }
  }

  return React.memo(GlobalStyleComponent);
}
