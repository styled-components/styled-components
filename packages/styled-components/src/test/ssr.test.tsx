/**
 * @jest-environment node
 */

import React from "react";
import { renderToString, renderToNodeStream } from "react-dom/server";
import stylisRTLPlugin from "stylis-plugin-rtl";
import ServerStyleSheet from "../models/ServerStyleSheet";
import { resetStyled } from "./utils";
import createGlobalStyle from "../constructors/createGlobalStyle";
import StyleSheetManager from "../models/StyleSheetManager";

jest.mock('../utils/nonce');

let styled: ReturnType<typeof resetStyled>;

describe('ssr', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});

    // eslint-disable-next-line
    require('../utils/nonce').mockReset();

    styled = resetStyled(true);
  });

  it('should extract the CSS in a simple case', () => {
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(sheet.collectStyles(<Heading>Hello SSR!</Heading>));
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should extract both global and local CSS', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(sheet.collectStyles(<React.Fragment>
          <Component />
          <Heading>Hello SSR!</Heading>
        </React.Fragment>));
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should not spill ServerStyleSheets into each other', () => {
    const A = styled.h1`
      color: red;
    `;
    const B = styled.h1`
      color: green;
    `;

    const sheetA = new ServerStyleSheet();
    renderToString(sheetA.collectStyles(<A />));
    const cssA = sheetA.getStyleTags();

    const sheetB = new ServerStyleSheet();
    renderToString(sheetB.collectStyles(<B />));
    const cssB = sheetB.getStyleTags();

    expect(cssA).toContain('red');
    expect(cssA).not.toContain('green');
    expect(cssB).not.toContain('red');
    expect(cssB).toContain('green');
  });

  it('should add a nonce to the stylesheet if webpack nonce is detected in the global scope', () => {
    // eslint-disable-next-line
    require('../utils/nonce').mockImplementation(() => 'foo');

    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(sheet.collectStyles(<React.Fragment>
          <Component />
          <Heading>Hello SSR!</Heading>
        </React.Fragment>));
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should render CSS in the order the components were defined, not rendered', () => {
    const ONE = styled.h1.withConfig({ componentId: 'ONE' })`
      color: red;
    `;
    const TWO = styled.h2.withConfig({ componentId: 'TWO' })`
      color: blue;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(sheet.collectStyles(<div>
          <TWO />
          <ONE />
        </div>));
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should return a generated React style element', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();

    renderToString(sheet.collectStyles(<React.Fragment>
          <Component />
          <Heading>Hello SSR!</Heading>
        </React.Fragment>));

    const [element] = sheet.getStyleElement();

    expect(element.props.dangerouslySetInnerHTML).toBeDefined();
    expect(element.props.children).not.toBeDefined();
    expect(element.props).toMatchSnapshot();
  });

  it('should return a generated React style element with nonce if webpack nonce is preset in the global scope', () => {
    // eslint-disable-next-line
    require('../utils/nonce').mockImplementation(() => 'foo');

    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();

    renderToString(sheet.collectStyles(<React.Fragment>
          <Heading>Hello SSR!</Heading>
          <Component />
        </React.Fragment>));

    const [element] = sheet.getStyleElement();
    expect(element.props.nonce).toBe('foo');
  });

  it('should interleave styles with rendered HTML when utilitizing streaming', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>);
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => {
        expect(received).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        resolve();
      });

      stream.on('error', reject);
    });
  });

  it('should interleave styles with rendered HTML when chunked streaming', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const Body = styled.div`
      color: blue;
    `;

    const SideBar = styled.div`
      color: yellow;
    `;

    const Footer = styled.div`
      color: green;
    `;

    // This is the result of the above
    const expectedElements = '<div>*************************</div>'.repeat(100);

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
        <Body>
          {new Array(1000).fill(0).map((_, i) => <div key={i}>*************************</div>)}
        </Body>
        <SideBar>SideBar</SideBar>
        <Footer>Footer</Footer>
      </React.Fragment>);

    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));
    const stream$ = new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => resolve(received));
      stream.on('error', reject);
    });

    return stream$.then(received => {
      expect(sheet.sealed).toBe(true);
      expect(received.includes(expectedElements)).toBeTruthy();
      expect(received).toMatch(/yellow/);
      expect(received).toMatch(/green/);
    });
  });

  it('should handle errors while streaming', () => {
    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(null);
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise(resolve => {
      stream.on('data', () => {});

      stream.on('error', err => {
        expect(err).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        resolve();
      });
    });
  });

  it('should not interleave style tags into textarea elements', () => {
    const StyledTextArea = styled.textarea`
      height: ${props => `${props.height}px`};
    `;

    const sheet = new ServerStyleSheet();

    // Currently we cannot set the chunk size to read with react renderToNodeStream, so to ensure
    // that multiple chunks are created, we initialize a large array of styled text areas.  We give
    // each textarea a different style to ensure a large enough number of style tags are generated
    // to be interleaved in the document
    const jsx = sheet.collectStyles(<React.Fragment>
        {new Array(500).fill(0).map((_, i) => <StyledTextArea key={i} className="test-textarea" onChange={() => {}} value={`Textarea ${i}`} height={i} />)}
      </React.Fragment>);

    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => {
        const styleTagsInsideTextarea = received.match(/<\/style>[^<]*<\/textarea>/g);

        expect(styleTagsInsideTextarea).toBeNull();
        resolve();
      });

      stream.on('error', reject);
    });
  });

  it('should throw if interleaveWithNodeStream is called twice', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>);

    expect(() => sheet.interleaveWithNodeStream(sheet.interleaveWithNodeStream(renderToNodeStream(jsx)))).toThrowErrorMatchingSnapshot();
  });

  it('should throw if getStyleTags is called after interleaveWithNodeStream is called', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();

    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>);

    sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    expect(sheet.getStyleTags).toThrowErrorMatchingSnapshot();
  });

  it('should throw if getStyleElement is called after interleaveWithNodeStream is called', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();

    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>);

    sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    expect(sheet.getStyleElement).toThrowErrorMatchingSnapshot();
  });

  it('should throw if getStyleTags is called after streaming is complete', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>);
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => {
        expect(received).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        expect(sheet.getStyleTags).toThrowErrorMatchingSnapshot();

        resolve();
      });

      stream.on('error', reject);
    });
  });

  it('should throw if getStyleElement is called after streaming is complete', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(<React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>);
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => {
        expect(received).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        expect(sheet.getStyleElement).toThrowErrorMatchingSnapshot();

        resolve();
      });

      stream.on('error', reject);
    });
  });

  it('should work with stylesheet manager and passed stylis plugins', () => {
    const Heading = styled.h1`
      padding-left: 5px;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(sheet.collectStyles(<StyleSheetManager stylisPlugins={[stylisRTLPlugin]}>
          <Heading>Hello SSR!</Heading>
        </StyleSheetManager>));
    const css = sheet.getStyleTags();

    expect(html).toMatchInlineSnapshot(`
      <h1 class="sc-a b">
        Hello SSR!
      </h1>
    `);
    expect(css).toMatchInlineSnapshot(`
      <style data-styled="true"
             data-styled-version="JEST_MOCK_VERSION"
      >
        .b{padding-right:5px;}/*!sc*/
      data-styled.g1[id="sc-a"]{content:"b,"}/*!sc*/
      </style>
    `);
  });
});