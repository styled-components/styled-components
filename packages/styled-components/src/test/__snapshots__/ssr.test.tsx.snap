// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr should add a nonce to the stylesheet if webpack nonce is detected in the global scope 1`] = `
<h1 class="sc-b c">
  Hello SSR!
</h1>
`;

exports[`ssr should add a nonce to the stylesheet if webpack nonce is detected in the global scope 2`] = `
<style nonce="foo"
       data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  body{background:papayawhip;}/*!sc*/
data-styled.g1[id="sc-global-a1"]{content:"sc-global-a1,"}/*!sc*/
.c{color:red;}/*!sc*/
data-styled.g2[id="sc-b"]{content:"c,"}/*!sc*/
</style>
`;

exports[`ssr should extract both global and local CSS 1`] = `
<h1 class="sc-b c">
  Hello SSR!
</h1>
`;

exports[`ssr should extract both global and local CSS 2`] = `
<style data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  body{background:papayawhip;}/*!sc*/
data-styled.g1[id="sc-global-a1"]{content:"sc-global-a1,"}/*!sc*/
.c{color:red;}/*!sc*/
data-styled.g2[id="sc-b"]{content:"c,"}/*!sc*/
</style>
`;

exports[`ssr should extract the CSS in a simple case 1`] = `
<h1 class="sc-a b">
  Hello SSR!
</h1>
`;

exports[`ssr should extract the CSS in a simple case 2`] = `
<style data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  .b{color:red;}/*!sc*/
data-styled.g1[id="sc-a"]{content:"b,"}/*!sc*/
</style>
`;

exports[`ssr should handle errors while streaming 1`] = `[Error: React.Children.only expected to receive a single React element child.]`;

exports[`ssr should interleave styles with rendered HTML when utilitizing streaming 1`] = `
<style data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  body{background:papayawhip;}/*!sc*/
data-styled.g1[id="sc-global-a1"]{content:"sc-global-a1,"}/*!sc*/
.c{color:red;}/*!sc*/
data-styled.g2[id="sc-b"]{content:"c,"}/*!sc*/
</style>
<h1 class="sc-b c">
  Hello SSR!
</h1>
`;

exports[`ssr should render CSS in the order the components were defined, not rendered 1`] = `
<div>
  <h2 class="TWO a">
  </h2>
  <h1 class="ONE b">
  </h1>
</div>
`;

exports[`ssr should render CSS in the order the components were defined, not rendered 2`] = `
<style data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  .b{color:red;}/*!sc*/
data-styled.g1[id="ONE"]{content:"b,"}/*!sc*/
.a{color:blue;}/*!sc*/
data-styled.g2[id="TWO"]{content:"a,"}/*!sc*/
</style>
`;

exports[`ssr should return a generated React style element 1`] = `
Object {
  "dangerouslySetInnerHTML": Object {
    "__html": "body{background:papayawhip;}/*!sc*/
data-styled.g1[id=\\"sc-global-a1\\"]{content:\\"sc-global-a1,\\"}/*!sc*/
.c{color:red;}/*!sc*/
data-styled.g2[id=\\"sc-b\\"]{content:\\"c,\\"}/*!sc*/
",
  },
  "data-styled": "",
  "data-styled-version": "JEST_MOCK_VERSION",
}
`;

exports[`ssr should throw if getStyleElement is called after interleaveWithNodeStream is called 1`] = `
"Can't collect styles once you've consumed a \`ServerStyleSheet\`'s styles! \`ServerStyleSheet\` is a one off instance for each server-side render cycle.

- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?"
`;

exports[`ssr should throw if getStyleElement is called after streaming is complete 1`] = `
<style data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  body{background:papayawhip;}/*!sc*/
data-styled.g1[id="sc-global-a1"]{content:"sc-global-a1,"}/*!sc*/
.c{color:red;}/*!sc*/
data-styled.g2[id="sc-b"]{content:"c,"}/*!sc*/
</style>
<h1 class="sc-b c">
  Hello SSR!
</h1>
`;

exports[`ssr should throw if getStyleElement is called after streaming is complete 2`] = `
"Can't collect styles once you've consumed a \`ServerStyleSheet\`'s styles! \`ServerStyleSheet\` is a one off instance for each server-side render cycle.

- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?"
`;

exports[`ssr should throw if getStyleTags is called after interleaveWithNodeStream is called 1`] = `
"Can't collect styles once you've consumed a \`ServerStyleSheet\`'s styles! \`ServerStyleSheet\` is a one off instance for each server-side render cycle.

- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?"
`;

exports[`ssr should throw if getStyleTags is called after streaming is complete 1`] = `
<style data-styled="true"
       data-styled-version="JEST_MOCK_VERSION"
>
  body{background:papayawhip;}/*!sc*/
data-styled.g1[id="sc-global-a1"]{content:"sc-global-a1,"}/*!sc*/
.c{color:red;}/*!sc*/
data-styled.g2[id="sc-b"]{content:"c,"}/*!sc*/
</style>
<h1 class="sc-b c">
  Hello SSR!
</h1>
`;

exports[`ssr should throw if getStyleTags is called after streaming is complete 2`] = `
"Can't collect styles once you've consumed a \`ServerStyleSheet\`'s styles! \`ServerStyleSheet\` is a one off instance for each server-side render cycle.

- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?"
`;

exports[`ssr should throw if interleaveWithNodeStream is called twice 1`] = `
"Can't collect styles once you've consumed a \`ServerStyleSheet\`'s styles! \`ServerStyleSheet\` is a one off instance for each server-side render cycle.

- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?"
`;
